cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -269,114 +285,690 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+		return fmt.Errorf("cannot validate boot: %s mountpoint is expected to be from disk %s but is not", name, m.disk.Dev())
+	}
+	return nil
+}
+
+// ensureUnlockResConsistency does a couple of things for ubuntu-data and
+// ubuntu-save respectively, it:
+// - updates the state machine with new, consistent information from the unlock
+//   result
+// - ensures that previous state the state machine has seen is consistent with
+//   the new unlock result we have
+// - verifies the logical consistency of the unlock error and the unlock result
+func (m *stateMachine) ensureUnlockResConsistency(part string, unlockRes secboot.UnlockResult, unlockErr error) error {
+	var stateDevice *string
+	switch part {
+	case "ubuntu-data":
+		stateDevice = &m.dataDevice
+	case "ubuntu-save":
+		stateDevice = &m.saveDevice
+	}
+	// TODO: we do a lot of checking on the result here to simplify later
+	//       decisions, perhaps this is too much?
+
+	// before checking err, make sure that the results from UnlockRes match
+	// what we've seen before and put valid results inside the state machine
+
+	// ensure consistency between encrypted state of the device/disk and what we
+	// may have seen previously
+	if m.isEncryptedDev && !unlockRes.IsDecryptedDevice {
+		// then we previously were able to positively identify an
+		// ubuntu-data-enc but can't anymore, so we have inconsistent results
+		// from inspecting the disk which is suspicious and we should fail
+		return fmt.Errorf("inconsistent disk encryption status: previous access resulted in encrypted, but now is unencrypted from partition %s", part)
+	}
+
+	// if isEncryptedDev hasn't been set on the state machine yet, then set that
+	// on the state machine before continuing - this is okay because we might
+	// not have been able to do anything with ubuntu-data if we couldn't mount
+	// ubuntu-boot, so this might be the first time we tried to unlock
+	// ubuntu-data and m.isEncryptedDev may have the default value of false
+	if !m.isEncryptedDev && unlockRes.IsDecryptedDevice {
+		m.isEncryptedDev = unlockRes.IsDecryptedDevice
+	}
+
+	// also make sure that if we previously saw a device that we see the same
+	// device again
+	if unlockRes.Device != "" && *stateDevice != "" && unlockRes.Device != *stateDevice {
+		return fmt.Errorf("inconsistent partitions found for %s: previously found %s but now found %s", part, *stateDevice, unlockRes.Device)
+	}
+
+	if unlockRes.Device != "" {
+		*stateDevice = unlockRes.Device
+		m.degradedState.partition(part).LocateState = partitionFound
+		m.degradedState.partition(part).Device = unlockRes.Device
+	}
+
+	if !unlockRes.IsDecryptedDevice && unlockRes.Device != "" && unlockErr != nil {
+		// this case should be impossible to enter, if we have an unencrypted
+		// device and we know what the device is then what is the error?
+		return fmt.Errorf("internal error: inconsistent return values from UnlockVolumeUsingSealedKeyIfEncrypted for partition %s", part)
+	}
+
+	return nil
+}
```

Comment:
I wonder if the code would be a bit less repetitive and easier to follow if either partitionState or stateMachine had methods like:

setMountState(part, err)
setLocateState(part, err)
setUnlockStateRunKey(part, UnlockRes, err)
setUnlockStateFallbackKey(part, Unlockres, err)

or similar ?

Reply:
Adding those helper was nice, thanks for that!